### PR TITLE
feat(refunds): update refunds filters

### LINKED ISSUE #4408: update refunds filters and list
Include filter for amount and merchant connector id in refunds list.

### PR DESCRIPTION
**Pull request by apoorvdixit88**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
The PR:
- Adds support for new `filter_v2` api to get list of available filters for refunds quickly: connector with their corresponding label with merchant_connector_ids, currency and refund status. This api is quicker than previous one and gives static and dynamic filter. In this list of available value doesn't depend on refunds that are made (present in table). But these are static and configuration dependent.
- Modifies existing refunds list as per filters applied (also manages pagination result etc), we can filter list by amount and merchant connector ids as well and the application of previous filters on list remains same.

- _Also changed payments filter v2 (it is not being used as of now)_ .

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Closes #4408 

## How did you test it?
To get list of filters:
```
curl --location 'http://localhost:8080/refunds/filter_v2' \
--header 'Content-Type: application/json' \
--header 'Cookie: token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMWZjNzgyOTUtM2JiMy00OGJjLThlZDgtNzFjMzVjMzMxYWU2IiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNjk3NzE2ODcwIiwicm9sZV9pZCI6Im1lcmNoYW50X2FkbWluIiwiZXhwIjoxNjk3ODg5NjcxLCJvcmdfaWQiOiJvcmdfejQ5ZExMeTdmbllUODN5TDY3clEifQ.CJzEQ2qbhn-qUiiVBSdvCJiLvWp-5wCF9R54gth6QbQ' \
--header 'Authorization: Bearer JWT' \
--data ''
```

Response:
```
{
    "connector": {
"paypal_test": [
            {
                "connector_label": "payal_test_default",
                "merchant_connector_id": "mca_test"
            }
},
    "currency": [
        "AED",
        "ALL",
        "AMD",
        "ANG",
        "AOA",
        "ARS",
        "AUD",
        "AWG",
        "AZN",
        "BAM",
        "BBD",
        "BDT",
        "BGN",
        "BHD",
        "BIF",
        "BMD",
        "BND",
        "BOB",
        "BRL",
        "BSD",
        "BWP",
        "BYN",
        "BZD",
        "CAD",
        "CHF",
        "CLP",
        "CNY",
        "COP",
        "CRC",
        "CUP",
        "CVE",
        "CZK",
        "DJF",
        "DKK",
        "DOP",
        "DZD",
        "EGP",
        "ETB",
        "EUR",
        "FJD",
        "FKP",
        "GBP",
        "GEL",
        "GHS",
        "GIP",
        "GMD",
        "GNF",
        "GTQ",
        "GYD",
        "HKD",
        "HNL",
        "HRK",
        "HTG",
        "HUF",
        "IDR",
        "ILS",
        "INR",
        "IQD",
        "JMD",
        "JOD",
        "JPY",
        "KES",
        "KGS",
        "KHR",
        "KMF",
        "KRW",
        "KWD",
        "KYD",
        "KZT",
        "LAK",
        "LBP",
        "LKR",
        "LRD",
        "LSL",
        "LYD",
        "MAD",
        "MDL",
        "MGA",
        "MKD",
        "MMK",
        "MNT",
        "MOP",
        "MRU",
        "MUR",
        "MVR",
        "MWK",
        "MXN",
        "MYR",
        "MZN",
        "NAD",
        "NGN",
        "NIO",
        "NOK",
        "NPR",
        "NZD",
        "OMR",
        "PAB",
        "PEN",
        "PGK",
        "PHP",
        "PKR",
        "PLN",
        "PYG",
        "QAR",
        "RON",
        "RSD",
        "RUB",
        "RWF",
        "SAR",
        "SBD",
        "SCR",
        "SEK",
        "SGD",
        "SHP",
        "SLE",
        "SLL",
        "SOS",
        "SRD",
        "SSP",
        "STN",
        "SVC",
        "SZL",
        "THB",
        "TND",
        "TOP",
        "TRY",
        "TTD",
        "TWD",
        "TZS",
        "UAH",
        "UGX",
        "USD",
        "UYU",
        "UZS",
        "VES",
        "VND",
        "VUV",
        "WST",
        "XAF",
        "XCD",
        "XOF",
        "XPF",
        "YER",
        "ZAR",
        "ZMW"
    ],
    "refund_status": [
        "Failure",
        "ManualReview",
        "Pending",
        "Success",
        "TransactionFailure"
    ]
}
```

To filter refunds list by new parameters i.e., amount and merchant_connector_id
```
curl --location 'http://localhost:8080/refunds/list' \
--header 'Content-Type: application/json' \
--header 'Cookie: token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMWZjNzgyOTUtM2JiMy00OGJjLThlZDgtNzFjMzVjMzMxYWU2IiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNjk3NzE2ODcwIiwicm9sZV9pZCI6Im1lcmNoYW50X2FkbWluIiwiZXhwIjoxNjk3ODg5NjcxLCJvcmdfaWQiOiJvcmdfejQ5ZExMeTdmbllUODN5TDY3clEifQ.CJzEQ2qbhn-qUiiVBSdvCJiLvWp-5wCF9R54gth6QbQ' \
--header 'Authorization: Bearer JWT' \
--data '{
    "merchant_connector_id": ["mca_test1", "mca_test_2"],
   "connector": ["paypal_test"],
    "amount_filter": {
       "start_amount": 12600,
        "end_amount": 12600
    }
    
}'
```

Response:
```
{
    "count": 1,
    "total_count": 1,
    "data": [
        {
            "refund_id": "test_YqXmZZAtH9Hr97ZVA2lo",
            "payment_id": "test_KuyhzmNelel8VKEVLHUi",
            "amount": 12600,
            "currency": "USD",
            "status": "succeeded",
            "reason": "Sample Refund",
            "metadata": null,
            "error_message": null,
            "error_code": null,
            "created_at": "2024-04-22T05:58:24.000Z",
            "updated_at": "2024-04-22T05:58:24.000Z",
            "connector": "paypal_test",
            "profile_id": "pro_itP1UHdaBjqXsFF0WoAm",
            "merchant_connector_id": "mca_test1"
        }
    ]
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/routes/refunds.rs [hunk 1]
```diff
@@ -267,3 +268,36 @@ pub async fn refunds_filter_list(
...
 }
+
+/// Refunds - Filter V2
+///
+/// To list the refunds filters associated with list of connectors, currencies and payment statuses
+#[utoipa::path(
+    get,
+    path = "/refunds/filter_v2",
+    responses(
+        (status = 200, description = "List of static filters", body = RefundListMetaData),
```
Comment (ThisIsMani) [inline crates/router/src/routes/refunds.rs:279]: Is the response body type correct?

**Review by ThisIsMani — DISMISSED** [review]

File: crates/api_models/src/refunds.rs [hunk 2]
```diff
@@ -181,6 +190,18 @@ pub struct RefundListMetaData {
     pub refund_status: Vec<enums::RefundStatus>,
 }
 
+#[derive(Clone, Debug, serde::Serialize, ToSchema)]
+pub struct RefundListFilters {
+    /// The list of available connector filters
+    pub connector: HashMap<String, Vec<MerchantConnectorInfo>>,
```
Comment (hrithikesh026) [inline crates/api_models/src/refunds.rs:196]: What will be the key in this hashmap?

File: crates/router/src/core/refunds.rs [hunk 3]
```diff
@@ -767,6 +774,53 @@ pub async fn refund_filter_list(
...
+                    };
+                    (merchant_connector_account.connector_name.clone(), info)
+                })
+        })
+        .for_each(|(connector_name, info)| {
+            connector_map
+                .entry(connector_name.clone())
+                .or_default()
+                .push(info);
+        });
```
Comment (hrithikesh026) [inline crates/router/src/core/refunds.rs:813]:
```suggestion
    let mut connector_map: HashMap<String, Vec<MerchantConnectorInfo>> = 
    merchant_connector_accounts
        .iter()
        .filter_map(|merchant_connector_account| {
            merchant_connector_account
                .connector_label
                .as_ref()
                .map(|label| {
                    let info = MerchantConnectorInfo {
                        connector_label: label.clone(),
                        merchant_connector_id: merchant_connector_account
                            .merchant_connector_id
                            .clone(),
                    };
                    (merchant_connector_account.connector_name.clone(), info)
                })
        }).collect();
```

File: crates/api_models/src/refunds.rs (hunk 2 above)
Comment (hrithikesh026) [reply crates/api_models/src/refunds.rs:196]: Please mention it in the doc comment.

File: crates/router/src/core/refunds.rs (hunk 3 above)
Comment (hrithikesh026) [inline crates/router/src/core/refunds.rs:813]: fold() and collect() can be used here.
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold

**Review by hrithikesh026 — DISMISSED** [review]

**Review by ThisIsMani — DISMISSED** [review]

**Review by racnan — APPROVED** [review]
minor stuff

**Review by hrithikesh026 — APPROVED** [review]

**Review by racnan — DISMISSED** [review]

**Review by racnan — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Pull request merged**